src/test/ui/rfc-2497-if-let-chains/ensure-that-let-else-does-not-interact-with-let-chains.stderr

```diff
@@ -120,6 +126,12 @@ LL |     let Some(n) = opt && n == 1 else {
 error[E0308]: mismatched types
   --> $DIR/ensure-that-let-else-does-not-interact-with-let-chains.rs:15:19
    |
+LL |     let opt = Some(1i32);
+   |               ---------- here the type of `opt` is inferred to be `Option<_>`
+LL |
+LL |     let Some(n) = opt else {
+   |                   --- here the type of `opt` is inferred to be `Option<i32>`
```

Comment:
Hm, this doesn't look correct :thinking:

Reply:
To be more specific, the problem here is that we record the return type of the `Some` constructor before we unify the arguments.

Try adding a `resolve_vars_if_possible` to the type we pass to `write_ty` in [`check_expr_with_expectation_and_args`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_hir_typeck/expr.rs.html#264). That might help!